### PR TITLE
Refactor descriptor scans and build trees async per package

### PR DESCRIPTION
**Pull request by attiasas**
- [ ] All [tests](https://github.com/jfrog/jfrog-vscode-extension#building-and-testing-the-sources) passed. If this feature is not already covered by the tests, I added new tests.
- [x] I used `npm run format` for formatting the code before submitting the pull request.
-----
Changes:
* Remove redundant args for build tree
* Refactor scan specific methods from IssuesTreeDataProvider to Utils and loading to cache
* Make building tree async tasks to speed up the scan and make it more generalize to help maintenance 
* Refactor CI classes to separated folders and tidy treeProvider folder
* Make load task a background to notify users
* Fix bug for graph scan where componentId not exists (happened for some Yarn descriptors)
* Fix duplicated Python Environment objects where exists multiple descriptors in the same workspace
* Remove leftover comments and fix spell mistakes

### CODE REVIEW

File: src/main/treeDataProviders/utils/dependencyUtils.ts [hunk 1]
```diff
@@ -15,19 +15,288 @@ import { YarnUtils } from '../../utils/yarnUtils';
...
+        // Dependency graph scan task
+        await DependencyUtils.scanProjectDependencyGraph(
+            scanManager,
+            <DependencyScanResults>fileScanBundle.data,
+            fileScanBundle.dataNode,
+            rootGraph,
+            scanProgress,
+            scanProgress.onProgress
+        )
+            .then(issuesFound => {
```
Comment (yahavi) [inline src/main/treeDataProviders/utils/dependencyUtils.ts:188]:
```suggestion
            .then((issuesFound: number) => {
```

File: src/main/treeDataProviders/utils/dependencyUtils.ts [hunk 2]
```diff
@@ -164,6 +402,37 @@ export class DependencyUtils {
         return projectNode.dependenciesWithIssue.length;
     }
 
+    /**
+     * Create an Issue node child for a dependency issues node base on a given IVulnerability
+     * @param issue - the issue to create the child node from
+     * @param dependencyWithIssue - the parent of this issue
+     * @param severity - the severity of this issue
```
Comment (yahavi) [inline src/main/treeDataProviders/utils/dependencyUtils.ts:409]: Missing explanation about "component" argument

File: src/main/treeDataProviders/utils/dependencyUtils.ts [hunk 3]
```diff
@@ -164,6 +402,37 @@ export class DependencyUtils {
...
+        dependencyWithIssue: DependencyIssuesTreeNode,
+        severity: Severity,
+        component: IComponent,
+        impactedPath: IImpactGraph
+    ) {
+        let violationIssue: IViolation = <IViolation>issue;
+        if (violationIssue && violationIssue.license_key) {
+            // License violation
+            dependencyWithIssue.issues.push(new LicenseIssueTreeNode(violationIssue, severity, dependencyWithIssue, impactedPath));
+        } else {
```
Comment (yahavi) [inline src/main/treeDataProviders/utils/dependencyUtils.ts:423]: Let's use early return in this function. if-else-if-else is too much to comprehend for the random reader.

**Review by yahavi — CHANGES_REQUESTED** [review]
Well done @attiasas!

File: src/main/treeDataProviders/dependenciesTree/dependenciesRoot/yarnTree.ts [hunk 4]
```diff
@@ -4,43 +4,36 @@ import { GeneralInfo } from '../../../types/generalInfo';
...
+            this._logManager.logMessageAndToastErr(
                 `Failed to scan Yarn project. Hint: Please make sure the command "yarn install" runs successfully in ` + this.workspaceFolder + '".',
                 'ERR'
             );
-            yarnListFailed = true;
+            this.buildError = BuildTreeErrorType.NotInstalled;
         }
-        if (!yarnListFailed) {
+        if (!this.buildError) {
             this.populateDependencyTree(this, listResults?.data?.trees);
```
[the remark this answers is not in the document]
Comment (attiasas) [reply src/main/treeDataProviders/dependenciesTree/dependenciesRoot/yarnTree.ts]: done

File: src/main/treeDataProviders/utils/analyzerUtils.ts [hunk 5]
```diff
@@ -93,6 +94,57 @@ export class AnalyzerUtils {
...
+        fileScanBundle: FileScanBundle,
+        abortController: AbortController
+    ): Promise<void> {
+        let cvesToScan: string[] = [];
+        if (!(fileScanBundle.dataNode instanceof DescriptorTreeNode)) {
+            return;
+        }
+        let descriptorIssues: DependencyScanResults = <DependencyScanResults>fileScanBundle.data;
+        fileScanBundle.dataNode.issues.forEach(issue => {
+            if (issue instanceof CveTreeNode && !issue.parent.indirect && issue.cve?.cve && !cvesToScan.includes(issue.cve?.cve)) {
```
[the remark this answers is not in the document]
Comment (attiasas) [reply src/main/treeDataProviders/utils/analyzerUtils.ts]: nope, changed to set

**Pull request merged**